### ISSUE TITLE
Too long time generating PDF 

### ISSUE THREAD

**Issue by freekbass** [barryvdh/laravel-dompdf#586]
I have a report with a table, which normally generates 2000 records

When I export to PDF, this generation can take up to 7 minutes and a memory consumption of almost 1GB, plus use almost 70% of processor usage in that period, and in the end, my pdf file is 150kb in size.

I'm using opcache to try to improve, but there is some process to improve this execution. I have the same query generated in csv, which takes 30 seconds to generate.

I use the following code to generate the report

```
 $result = $this->agruparListaEntregasPendentes($entregas);
        $view = PDF::loadView('reports.relatorio_entrega_pendente.index',
            [
                'listaEntregasAgrupada' => $result,
                'filter' => $request,
                'custoTotal' => $custoTotal,
                'quantidadeTotal' => $quantidadeTotal,
                'cabecalhos' => $request->all(),
            ]);

        return $view->stream('Relatorio_Entrega_Pendente_' . getUniqueName() . '.pdf');
```

And on my blade, I have a nested foreach, but the return is grouped before, to make it easier.

```
@foreach($listaEntregasAgrupada as $campoAgrupar => $dadosMostrar)
            <tr class="group1">
                <td colspan="6">
                    <b>{{trans('words.label.' . ($filter->agrupador == 'S' ? "setor" : ($filter->agrupador == 'C' ? "cargo" : ($filter->agrupador == 'G' ? "ghe" : "centro_custo"))))}}
                        : {{!$campoAgrupar ? trans('words.label.sem_registro') : $campoAgrupar}}</b>
                </td>
            </tr>
            @foreach($dadosMostrar as $campoMostrar => $dadosDetalhes)
                <tr class="group2">
                    <td colspan="6">
                        <b>{{trans('words.label.' . ($filter->mostrar_por == 'P' ? ($filter->mostrar_grupo_produto == 'S' ? "grupo_produto" : "produto") : "funcionario"))}}
                            : {{!$campoMostrar ? trans('words.label.sem_registro') : $campoMostrar}}</b>
                    </td>
                </tr>
                @foreach($dadosDetalhes as $detalhe)
                    <tr>
                        <td>{{$detalhe->codigo_detalhe}}</td>
                        <td>{{$filter->mostrar_por == 'F' && $filter->mostrar_grupo_produto == 'S' && !$detalhe->descricao_detalhe ? trans('words.label.sem_grupo') : $detalhe->descricao_detalhe}}</td>
                        @if(!$detalhe->dt_entrega)
                            <td colspan="2"><span
                                        class="txtNuncaEntregue">{{trans('words.label.produto_nunca_entregue')}}</span>
                            </td>
                            <td>
                                <div class="alinhamento-direita">{{number_format($detalhe->vl_custo, 2, ',', '.')}}</div>
                            </td>
                            <td></td>
                        @else
                            <td>{{\Carbon\Carbon::parse($detalhe->dt_entrega)->format('d/m/Y')}}</td>
                            <td>{{\Carbon\Carbon::parse($detalhe->dt_previsao)->format('d/m/Y')}}</td>
                            <td>
                                <div class="alinhamento-direita">{{number_format($detalhe->vl_custo, 2, ',', '.')}}</div>
                            </td>
                            <td>
                                <div class="alinhamento-direita">
                                    {{$detalhe->nr_dias_atraso}}
                                </div>
                            </td>
                        @endif
                    </tr>
                @endforeach
            @endforeach
        @endforeach
```

Is this a bug in the library or is it possible to optimize?
Tks for help.

**Comment by syond** [comment]
Ta a mesma coisa aqui. Passei 4 variáveis por compact e a pagina fica carregando em loop e não gera nada.

**Comment by cristianmarint** [comment]
Having a related issue... Does anyone know if the library suppor loops inside the blade?

**Comment by Ashrafdev** [comment]
Solution: use wkhtmltopdf https://github.com/barryvdh/laravel-snappy

**Comment by QamarRafhan** [comment]
Facing same issue any solution?

**Comment by Nuhel** [comment]
I am also having this issue, I have no remote image or css

**Comment by vanwhebin** [comment]
same here 
it takes 45s to render pdf, which only 195k with 6 pics

**Comment by SHJordan** [comment]
Things to have in mind:
Enable opcache.
Install ImageMagick and enable it.
Do not run with xdebug enabled. If you use a IDE such as PHPStorm there is 0 need to enable it on .ini as you can simply load it on demand.
ini "okay" settings:
memory_limit = 512M
post_max_size = 512M
upload_max_filesize = 256M

**Comment by acgtwentyone** [comment]
Same issues. It takes almost five minutes to render a simple pdf with no image but external css. It seems like it has issues if i link external css.

// takes too long to render
```
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8">
    <title>SINFRA - Test Dom Pdf </title>
    <link rel="icon" href="{{ URL::to('/') }}/images/favicon.png" type="image/x-icon">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    {{-- <link rel="stylesheet" href="{{ URL::to('/') }}/css/pdf.css"> --}}  // here
    <link rel="stylesheet" href="{{ URL::to('/') }}/css/site/bootstrap2_pdf.css">  // here

</head>
<section>
    <div>
        <span>{{ date("Y-m-d H:m:s") }}</span>
    </div>
</section>
</body>

</html>
```

// render fast
```
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8">
    <title>SINFRA - Test Dom Pdf </title>
    <link rel="icon" href="{{ URL::to('/') }}/images/favicon.png" type="image/x-icon">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    {{-- <link rel="stylesheet" href="{{ URL::to('/') }}/css/pdf.css"> --}}  // here
    {{-- <link rel="stylesheet" href="{{ URL::to('/') }}/css/site/bootstrap2_pdf.css"> --}}  // here

</head>
<section>
    <div>
        <span>{{ date("Y-m-d H:m:s") }}</span>
    </div>
</section>
</body>

</html>
```

**Comment by dmcbrn** [comment]
Beware of https://github.com/dompdf/dompdf/issues/1326

Rendering long tables takes really long time. I don't se a working solution for this.

I believe these kind of pitfalls should be listed at the top of the README file.

**Comment by SHJordan** [comment]
> Beware of [dompdf/dompdf#1326](https://github.com/dompdf/dompdf/issues/1326)
> 
> Rendering long tables takes really long time. I don't se a working solution for this.
> 
> I believe these kind of pitfalls should be listed at the top of the README file.

Any suggestion on alternatives using PHP/Laravel/Inertia/Vue?

**Comment by dmcbrn** [comment]
This lib [wkhtmltopdf/wkhtmltopdf](https://github.com/wkhtmltopdf/wkhtmltopdf) looks most promising but I didn't have the time to play with it.

It needs to be installed to your server manually as it seems it's a C++ library so it's a bit of extra work with installation and maintenance. 

Then, I suppose you use something like [barryvdh/laravel-snappy](https://github.com/barryvdh/laravel-snappy) to use it.

There's some info that could be helpful here as well: https://sternerstuff.dev/2021/06/configure-wkhtmltopdf-for-laravel-snappy-on-homestead-sail-forge/

Feel free to respond if you check it out.

**Comment by chandachewe10** [comment]
Make sure you have no external CSS, if you have to use one try by all means to use internal CSS and avoid external bootstrap links together with images. These things enables domPDF to load slowly

**Comment by raisurrehman-trs** [comment]
I'm using DomPDF in a Laravel app to generate a PDF that includes 138 items (simple table rows). The view loads in about 72 seconds, but the PDF generation (->download()) takes over 6 minutes.
I also tried Snappy (wkhtmltopdf), but it was even slower (around 10 minutes).
I've already:
Increased memory and execution time
Still, the PDF takes too long to generate. Is this normal? Are there any best practices to speed this up?
Thanks!

**Comment by RH203** [comment]
@raisurrehman-trs In my case, I need to generate around 16,000 rows, and I ran into an out-of-memory exception when using Dompdf.

What worked for me was returning the data as JSON from the backend, then generating the PDF on the client side using [jsPDF](https://www.npmjs.com/package/jspdf).

I went with this approach because the client devices in my case are powerful enough to handle the process. Of course, this might not be a good fit if the users are on low-end devices, since the memory and processing load is basically moved from the server to the client.

I also found [this comment](https://github.com/barryvdh/laravel-dompdf/issues/77#issuecomment-65786370)(https://github.com/barryvdh/laravel-dompdf/issues/77#issuecomment-65786370) mentioning that Dompdf can use a huge amount of memory when rendering long HTML documents, so this seems to be a known issue.

I’m not sure whether @barryvdh can fix this directly, since the issue seems to come from Dompdf itself. Anyone interested in helping can contribute to the main project here: [dompdf/dompdf](https://github.com/dompdf/dompdf).
